platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java

```diff
@@ -112,99 +122,113 @@ private File getJavaHome(File markedLocation) {
-        final File[] installFolder = new File[1];
+    public File provisionFromArchive(JavaToolchainSpec spec, File jdkArchive, URI uri) throws IOException {
+        // Unpack into temporary directory (but on same file system as our target directory location)
+        File unpackFolder = unpack(jdkArchive);
         try {
-            unpackFolder[0] = unpack(jdkArchive);
-
-            //put the marker file in the unpack folder from where it will get in its proper place
-            //when the contents are copied to the installation folder
-            File markedLocation = markLocationInsideFolder(unpackFolder[0]);
+            // Get the folder that is the real root of the unpacked JDK, skipping any archive root folder
+            File unpackedRoot = determineUnpackedRoot(unpackFolder);
 
-            //probe unpacked installation for its metadata
-            JvmInstallationMetadata metadata = getMetadata(markedLocation);
+            // Probe unpacked installation for its metadata
+            JvmInstallationMetadata metadata = getMetadata(unpackedRoot);
 
             validateMetadataMatchesSpec(spec, uri, metadata);
 
-            String installFolderName = getInstallFolderName(metadata);
-            installFolder[0] = new File(jdkDirectory, installFolderName);
-
-            //make sure the install folder is empty
-            checkInstallFolderForLeftoverContent(installFolder[0], uri, spec, metadata);
-            operations.delete(installFolder[0]);
-
-            //copy content of unpack folder to install folder, including the marker file
-            operations.copy(copySpec -> {
-                copySpec.from(unpackFolder[0]);
-                copySpec.into(installFolder[0]);
-            });
-
-            LOGGER.info("Installed toolchain from {} into {}", uri, installFolder[0]);
-            return getJavaHome(markedLocation(installFolder[0]));
-        } catch (Throwable t) {
-            // provisioning failed, clean up leftovers
-            if (installFolder[0] != null) {
-                operations.delete(installFolder[0]);
+            // Check our target directory, to see if anything exists there, and if so, is it marked as ready?
+            File installFolder = new File(jdkDirectory, getInstallFolderName(metadata));
+            if (!installFolder.getParentFile().mkdirs() && !installFolder.getParentFile().isDirectory()) {
+                throw new IOException("Failed to create install parent directory: " + installFolder.getParentFile());
             }
-            throw t;
+            // Before checking existence, lock the install folder name to prevent concurrent installations
+            try (FileLock ignored = acquireWriteLock(installFolder, "Provisioning JDK from " + uri)) {
+                if (installFolder.exists()) {
+                    if (isMarkedLocation(installFolder)) {
+                        LOGGER.info("Toolchain from {} already installed at {}", uri, installFolder);
```

Comment:
🤔  I feel like we should be making this check before we even download and unpack the archive at all. Granted, I accept that might be too much of a change for this PR and be out of scope, but it seems we should verify that we don't need to download the JDK before downloading the JDK archive

Reply:
In most cases this check is already done (logic before this scans the JDKs folder for installed options); this is a fallback case for concurrent provisioning.